### ISSUE TITLE
Where-clause for bindings

### ISSUE THREAD

**Issue by keddelzz** [morganey-lang/Morganey#111]
Introduce `where`-clause for bindings for eliminating duplicate code.

Instead of writing

```
Y := (λg . ((λx . (g (x x))) (λx . (g (x x)))))
```

you could write

```
Y := (λg . (gid gid))
  where {
    gid := (λx . (g (x x)))
  }
```
## 

As for the syntax: I introduced the `where`-keyword in the example, which uses curly-brackets. Currently Morganeys parser won't support the `where` clause, in which all local bindings are indented. (Because the parser doesn't care about whiteplace. If all bindings of a where should be determined by indentation, a rewrite of the parser might be necessary.)

I'm also fine with a different syntax.

**Comment by rexim** [comment]
I'm not sure about this one... Where-clause is essentially an idea of private bindings. Do we really need private bindings right now? I could rewrite the above expression like:

```
gid := (λx . (g (x x)))
Y := (λg . (gid gid))
```

And technically where-clause would be a change in the binding wrapping order (See #109), which is undefined right now.

I'm not against this proposal either. Just not sure if we need it now. Maybe a bit later once #109  is closed it will be more clear how to deal with this issue. Until then let's keep it a proposal. Thanks for filing it. :)

**Comment by keddelzz** [comment]
You are right, you could rewrite the above expression, but `gid` would in this case pollute the namespace. I think Morganey should have a form of private bindings: Either those implemented with the `where`-clause or private in the module (as shown in the example below):

```
private gid := (λx . (g (x x)))
Y := (λg . (gid gid))
```

Personally, I like the `where` clauses more.

**Comment by rexim** [comment]
I also like where-clause more. As I already said I'm not against this feature. I simply do not see how it fits the current semantic and execution of Morganey program. Furthermore, I don't see how any kind of access control fits there right now. What I suggest is to resolve #109 and then get back to this discussion. :)

I completely aware of namespace pollution with my approach. I just showed how we can deal with expression complexity right now, until it's clear how to fit any kind of access control (local bindings, private bindings, anything) into the semantic of Morganey.

**Comment by rexim** [comment]
Ok, I accept this proposal. After #109 it is clear how to implement that properly. Where-bindings are just another edges in the binding dependency graph that we already handle properly.

But this scope of this issue is too big. I suggest to narrow it down to updating `MorganeyBinding` AST node to support the where-clauses and implementing the corresponding semantic in `MorganeyCompiler`. For the syntax I'd suggest to create a separate issue.
